tools/istio-iptables/pkg/capture/run.go

```diff
@@ -359,6 +359,10 @@ func (cfg *IptablesConfigurator) Run() {
 	cfg.iptables.AppendRule(iptableslog.UndefinedCommand, constants.ISTIOINBOUND, constants.NAT, "-p", constants.TCP, "--dport",
 		cfg.cfg.InboundTunnelPort, "-j", constants.RETURN)
 
+	// Create a rule for invalid drop
+	cfg.iptables.AppendRule(iptableslog.UndefinedCommand, constants.ISTIOINBOUND, constants.MANGLE, "-m", "conntrack", "--ctstate",
```

Comment:
Why MANGLE? AFAIK we only use it for tproxy today?

Reply:
Is it because it is the first table that the rule will be dropped as early as possible

Reply:
Per my understanding, we need to configure iptables to drop invalid packets instead of forwarding them without address translation. Create it in INPUT chain would also help, I'll verify it in our case first to see if the reset is gone after this rule configured.
```
iptables -I ISTIOINBOUND -m conntrack --ctstate INVALID -j DROP
```

Reply:
The drop decision can be made anytime after the conntrack. Now that table MANGLE is the first after the conntrack, it's the indeed the best table

Reply:
@lambdai  Based on below diagram, I think we can configure the invalid drop in PREROUTING chain. 
![proxy](https://user-images.githubusercontent.com/12857255/147072652-f22d59a3-88f7-481a-8eb3-12fd10ff8cad.png)

Actually I tired to add to ISTIO_INBOUND but failed with error:
```
iptables -t mangle -A  ISTIO_INBOUND -p TCP    -m conntrack --ctstate INVALID -j DROP
iptables: No chain/target/match by that name.
```
And the reason seems like the ISTIO related chains are not in `mangle` table and `INPUT` chain won't work since only non-TCP traffic will go through it as per above diagram. 

```
iptables -t mangle  -L --line-numbers -n
Chain PREROUTING (policy ACCEPT)
num  target     prot opt source               destination
1    DROP       all  --  0.0.0.0/0            0.0.0.0/0            ctstate INVALID

Chain INPUT (policy ACCEPT)
num  target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
num  target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
num  target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
num  target     prot opt source               destination
```

After adding this rule I can see it now in the PREROUTING:
```
 iptables-save
# Generated by iptables-save v1.4.21 on Wed Dec 22 02:53:04 2021
*raw
:PREROUTING ACCEPT [25195:6428378]
:OUTPUT ACCEPT [22269:6017603]
COMMIT
# Completed on Wed Dec 22 02:53:04 2021
# Generated by iptables-save v1.4.21 on Wed Dec 22 02:53:04 2021
*mangle
:PREROUTING ACCEPT [23364:6031899]
:INPUT ACCEPT [23364:6031899]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [20556:5473772]
:POSTROUTING ACCEPT [20556:5473772]
**-A PREROUTING -m conntrack --ctstate INVALID -j DROP**
COMMIT
# Completed on Wed Dec 22 02:53:04 2021
# Generated by iptables-save v1.4.21 on Wed Dec 22 02:53:04 2021
*filter
:INPUT ACCEPT [65109:15777691]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [58256:15968553]
COMMIT
# Completed on Wed Dec 22 02:53:04 2021
# Generated by iptables-save v1.4.21 on Wed Dec 22 02:53:04 2021
*nat
:PREROUTING ACCEPT [358200:21495109]
:INPUT ACCEPT [374732:22297943]
:OUTPUT ACCEPT [830897:64816986]
:POSTROUTING ACCEPT [1520591:120009333]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A OUTPUT -p udp -m udp --dport 53 -m owner --uid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_INBOUND -p tcp -m tcp --dport 15008 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 9090 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 12020 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 8083 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -j ISTIO_IN_REDIRECT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -p tcp -m tcp --dport 443 -j RETURN
-A ISTIO_OUTPUT -p tcp -m tcp --dport 80 -j RETURN
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -p tcp -m tcp ! --dport 53 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -p tcp -m tcp ! --dport 53 -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -p tcp -m tcp ! --dport 53 -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -p tcp -m tcp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Wed Dec 22 02:53:04 2021
```

Reply:
BTW we are using `sidecar.istio.io/interceptionMode: REDIRECT`, seems in this mode the invalid drop it not supported in the ISTIO_INBOUND chain.

Reply:
The ISTIOINBOUND chain is in `nat` table with REDIRECT mode and the `nat` table is not intended for filtering, the use of DROP is therefore inhibited. So we'll need to add this rule in PREROUTING chain,
```
		if cfg.cfg.InboundInterceptionMode == constants.TPROXY {
			// When using TPROXY, create a new chain for routing all inbound traffic to
			// Envoy. Any packet entering this chain gets marked with the ${INBOUND_TPROXY_MARK} mark,
			// so that they get routed to the loopback interface in order to get redirected to Envoy.
			// In the ISTIOINBOUND chain, '-j ISTIODIVERT' reroutes to the loopback
			// interface.
			// Mark all inbound packets.
			cfg.iptables.AppendRule(iptableslog.UndefinedCommand, constants.ISTIODIVERT, constants.MANGLE, "-j", constants.MARK, "--set-mark",
				cfg.cfg.InboundTProxyMark)
			cfg.iptables.AppendRule(iptableslog.UndefinedCommand, constants.ISTIODIVERT, constants.MANGLE, "-j", constants.ACCEPT)

			// Create a new chain for redirecting inbound traffic to the common Envoy
			// port.
			// In the ISTIOINBOUND chain, '-j RETURN' bypasses Envoy and
			// '-j ISTIOTPROXY' redirects to Envoy.
			cfg.iptables.AppendVersionedRule("127.0.0.1/32", "::1/128", iptableslog.UndefinedCommand,
				constants.ISTIOTPROXY, constants.MANGLE, "!", "-d", constants.IPVersionSpecific,
				"-p", constants.TCP, "-j", constants.TPROXY,
				"--tproxy-mark", cfg.cfg.InboundTProxyMark+"/0xffffffff", "--on-port", cfg.cfg.InboundCapturePort)
			table = constants.MANGLE
		} else {
			table = constants.NAT
		}
		cfg.iptables.AppendRule(iptableslog.JumpInbound, constants.PREROUTING, table, "-p", constants.TCP,
			"-j", constants.ISTIOINBOUND)
```